src/v/rpc/test/rpc_gen_cycling_test.cc

```diff
@@ -337,15 +342,86 @@ FIXTURE_TEST(timeout_test, rpc_integration_fixture) {
 
     rpc::client<echo::echo_client_protocol> client(client_config());
     client.connect(model::no_timeout).get();
-    info("Calling echo method");
-    auto echo_resp = client.sleep_1s(
-      echo::echo_req{.str = "testing..."},
+    info("Calling sleep for.. 1s");
+    auto echo_resp = client.sleep_for(
+      echo::sleep_req{.secs = 1},
       rpc::client_opts(rpc::clock_type::now() + 100ms));
     BOOST_REQUIRE_EQUAL(
       echo_resp.get0().error(), rpc::errc::client_request_timeout);
     client.stop().get();
 }
 
+FIXTURE_TEST(timeout_test_cleanup_resources, rpc_integration_fixture) {
+    configure_server();
+    register_services();
+    start_server();
+
+    rpc::client<echo::echo_client_protocol> client(client_config());
+    client.connect(model::no_timeout).get();
+    using units_t = std::vector<ss::semaphore_units<>>;
+    ::mutex lock;
+    units_t units;
+    units.push_back(std::move(lock.get_units().get()));
+
+    auto opts = rpc::client_opts(rpc::clock_type::now() + 1s);
+    opts.resource_units = ss::make_foreign(
+      ss::make_lw_shared(std::move(units)));
+    ;
+
+    // Resources should now be owned by the client.
+    BOOST_REQUIRE(lock.try_get_units().has_value() == false);
+    auto echo_resp = client.sleep_for(
+      echo::sleep_req{.secs = 10}, std::move(opts));
+    BOOST_REQUIRE_EQUAL(
+      echo_resp.get0().error(), rpc::errc::client_request_timeout);
+    // Verify that the resource is released before the RPC finishes.
+    tests::cooperative_spin_wait_with_timeout(5s, [&] {
+        return lock.try_get_units().has_value();
+    }).get();
+    client.stop().get();
+}
+
+FIXTURE_TEST(test_flush_delays, rpc_integration_fixture) {
+    configure_server();
+    register_services();
+    start_server();
+
+    using units_t = std::vector<ss::semaphore_units<>>;
+    rpc::client<echo::echo_client_protocol> client(client_config());
+    client.connect(model::no_timeout).get();
+
+    constexpr int num_sends = 3000;
+    std::vector<::mutex> mutexes;
+    mutexes.reserve(num_sends);
+
+    // A bunch of RPCs with large sleep, this creates congestion.
```

Comment:
The test doesn't run into the problem deterministically (I only saw it happen a few times of many runs).. I  retained it because it does some sort of load tests with delayed RPCs..open to new ideas / removing this if seems unnecessary.

Reply:
Maybe it would be good to write a ducktape test where one of the nodes is isolated and we try to restart nodes that tries to communicate with the isolated one ?

Reply:
I added one.